src/diagonal.jl

```diff
@@ -970,6 +970,10 @@ function inv(D::Diagonal{T}) where T
     Diagonal(Di)
 end
 
+# Ensure doubly wrapped matrices use efficient diagonal methods and return a Symmetric/Hermitian type
+inv(A::Symmetric{<:Number,<:Diagonal}) = Symmetric(inv(A.data), sym_uplo(A.uplo))
+inv(A::Hermitian{<:Number,<:Diagonal}) = Hermitian(inv(real(A.data)), sym_uplo(A.uplo))
```

Comment:
To avoid the issue of returning a real-eltype matrix for a complex-eltype argument, this should perhaps be

```suggestion
inv(A::Hermitian{T,S}) where {T<:Number, S<:Diagonal} = Hermitian{T,S}(inv(real(A.data)), A.uplo)
```
?

Reply:
I suppose that's possible, but a Hermitian Diagonal matrix is actually always a real Diagonal matrix, so we can deduce from the type alone that the inverse is always a real diagonal matrix. This also holds for quaternion and octonion valued matrices. If there is a number type where becomes a problem, then using `real` might also be a problem.